### PR TITLE
RUN-2957 support all future multi-runtime events

### PR DESCRIPTION
**Pull request by whyn07m3**
ℹ️  Adds ability to subscribe to multi-runtime events for applications and windows that don't exist yet.

✅  [Test results look good](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/591c5c0a5114ac10ca6db3de)

---

➕  Added new ticket to add more tests
➕  New tests _(will work and will be published when multi-runtime is enabled by default)_:
1. [Application.addEventListener (initialized) (multi-runtime)](https://testing-dashboard.openfin.co/#/app/tests/5919b6615114ac10ca6db3bd/edit)
2. [Window.addEventListener (initialized) (multi-runtime)](https://testing-dashboard.openfin.co/#/app/tests/5919b3805114ac10ca6db3bb/edit)

### CODE REVIEW

**Comment by whyn07m3** [comment]
So I was just scolded by @datamadic for renaming the file in this PR and messing up diff'ing for that file, I admit I misbehaved and I won't do it anymore. Please, forgive me.

**Comment by HarsimranSingh** [comment]
@whyn07m3 rebase to develop and resolve the conflict plz

File: src/browser/remote_subscriptions.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            persistSubscription(subscription, runtime);
+            return;
+        }
+
+        // Unsubscribe in unneeded runtimes
+        unSubscribe(subscription, runtime);
+    });
+
+    // Remove subscription from the map of pending subscriptions
+    pendingRemoteSubscriptions.delete(subscription._id);
```
Comment (rdepena) [inline src/browser/remote_subscriptions.ts]: :neckbeard: assigning to null might be better.

File: src/browser/remote_subscriptions.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+/**
+ * Shape of a remote subscription
+ */
+interface IRemoteSubscription extends IRemoteSubscriptionProps {
+    _id: number; // ID of the subscription
+    isCleaned: boolean; // helps prevents repetitive un-subscriptions and other cleanup
+    unSubscriptions: { // a map of un-subscriptions assigned to runtime versions
+        [runtimeVersion: string]: (() => void)[]
```
Comment (rdepena) [inline src/browser/remote_subscriptions.ts]: this could be a Map.

File: src/browser/remote_subscriptions.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+/**
+ * Remove subscription from a given runtime
+ */
+function unSubscribe(subscription: IRemoteSubscription, runtime: PeerRuntime) {
+    const runtimeVersion = getRuntimeVersion(runtime);
+    const { unSubscriptions } = subscription;
+
+    unSubscriptions[runtimeVersion].forEach(unSubscribe => unSubscribe());
+    delete unSubscriptions[runtimeVersion];
```
Comment (rdepena) [inline src/browser/remote_subscriptions.ts]: same as above, null assignment but also not an issue if we just use a Map.

File: src/browser/remote_subscriptions.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+// all pending remote subscriptions are stored here
+const pendingRemoteSubscriptions: Map<number, IRemoteSubscription> = new Map();
+
+/**
+ * Shape of remote subscription props
+ */
+interface IRemoteSubscriptionProps {
+    uuid: string; // app uuid
+    name?: string; // window name
+    className: 'application'|'window'; // names of the class event emitters, used for subscriptions
```
Comment (whyn07m3) [inline src/browser/remote_subscriptions.ts]: @rdepena For external-application part, is [this](https://github.com/HadoukenIO/core/pull/20/files#diff-12aad93b38def9dae5259fe1abb6029bL183) handling it fully or not?

File: src/browser/remote_subscriptions.ts (hunk 1 above)
Comment (whyn07m3) [inline src/browser/remote_subscriptions.ts]: Let's keep `delete` so we don't enumerate over removed subscriptions

**Comment by whyn07m3** [comment]
✅  [Test results look good](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/591f53af5114ac10ca6db446)

File: src/browser/remote_subscriptions.ts (hunk 1 above)
Comment (rdepena) [inline src/browser/remote_subscriptions.ts]: As a moved along the file I saw that we should probably use a Map.

File: index.js [hunk 5]
```diff
@@ -56,9 +56,7 @@ import {
 
 import * as log from './src/browser/log';
 
-import {
-    applyAllPendingSubscriptions
-} from './src/browser/pending_subscriptions';
+import { applyAllRemoteSubscriptions } from './src/browser/remote_subscriptions';
```
Comment (HarsimranSingh) [inline index.js]: Did the lint rules change?

Comment (whyn07m3) [inline index.js]: I'll run through it again

**Comment by HarsimranSingh** [comment]
👍

**Pull request merged**